diff -rupN a/lib/bitmap.c b/lib/bitmap.c
--- a/lib/bitmap.c	2022-10-17 17:26:07.000000000 +0200
+++ b/lib/bitmap.c	2022-10-24 08:10:47.685459147 +0200
@@ -743,9 +743,8 @@ int bitmap_parse(const char *start, unsi
 	int chunks = BITS_TO_U32(nmaskbits);
 	u32 *bitmap = (u32 *)maskp;
 	int unset_bit;
-	int chunk;
 
-	for (chunk = 0; ; chunk++) {
+	while (1) {
 		end = bitmap_find_region_reverse(start, end);
 		if (start > end)
 			break;
@@ -753,11 +752,7 @@ int bitmap_parse(const char *start, unsi
 		if (!chunks--)
 			return -EOVERFLOW;
 
-#if defined(CONFIG_64BIT) && defined(__BIG_ENDIAN)
-		end = bitmap_get_x32_reverse(start, end, &bitmap[chunk ^ 1]);
-#else
-		end = bitmap_get_x32_reverse(start, end, &bitmap[chunk]);
-#endif
+		end = bitmap_get_x32_reverse(start, end, bitmap++);
 		if (IS_ERR(end))
 			return PTR_ERR(end);
 	}
