From 1989e0882fe90839d2e9188f07b801259d1719ab Mon Sep 17 00:00:00 2001
From: Thomas Zimmermann <tzimmermann@suse.de>
Date: Wed, 26 Aug 2026 11:03:11 +0200
Subject: [PATCH] mgag200: go looking for PInS in the video BIOS ROM on non-x86

---
 drivers/gpu/drm/mgag200/mgag200_g200.c | 24 +++++++++++++++++++++---
 1 file changed, 21 insertions(+), 3 deletions(-)

diff --git a/drivers/gpu/drm/mgag200/mgag200_g200.c b/drivers/gpu/drm/mgag200/mgag200_g200.c
index 9e6b4618fadd..159ec58a6541 100644
--- a/drivers/gpu/drm/mgag200/mgag200_g200.c
+++ b/drivers/gpu/drm/mgag200/mgag200_g200.c
@@ -239,7 +239,7 @@ static void mgag200_g200_interpret_bios(struct mgag200_g200_device *g200,
 	};
 	struct mga_device *mdev = &g200->base;
 	struct drm_device *dev = &mdev->base;
-	const unsigned char *pins;
+	const unsigned char *pins = NULL;
 	unsigned int pins_len, version;
 	int offset;
 	int tmp;
@@ -250,16 +250,34 @@ static void mgag200_g200_interpret_bios(struct mgag200_g200_device *g200,
 	if (memcmp(&bios[45], matrox, sizeof(matrox)) != 0)
 		return;
 
+#if defined(CONFIG_X86)
 	/* Get the PInS offset. */
 	if (size < MGA_BIOS_OFFSET + 2)
 		return;
 	offset = (bios[MGA_BIOS_OFFSET + 1] << 8) | bios[MGA_BIOS_OFFSET];
+	if (size < offset + 6)
+		return;
 
 	/* Get PInS data structure. */
+	pins = bios + offset;
+#else
+	/*
+	 * On OpenFirmware systems (PPC, MIPS, SPARC) the offset isn't stored
+	 * at the end of the BIOS image. Look for the PInS header instead.
+	 */
+	for (offset = 0 ; offset < size - 6; ++offset) {
+		const unsigned char *buf = bios + offset;
+
+		if (buf[0] == 0x2e && buf[1] == 0x41 && ((buf[2] == 64) || (buf[2] == 128))) {
+			pins = buf;
+			break;
+		}
+	}
 
-	if (size < offset + 6)
+	if (!pins)
 		return;
-	pins = bios + offset;
+#endif
+
 	if (pins[0] == 0x2e && pins[1] == 0x41) {
 		version = pins[5];
 		pins_len = pins[2];
-- 
2.55.0

