diff -rupN a/arch/powerpc/kvm/booke.c b/arch/powerpc/kvm/booke.c
--- a/arch/powerpc/kvm/booke.c	2026-09-20 22:48:15.000000000 +0200
+++ b/arch/powerpc/kvm/booke.c	2026-09-22 07:06:50.123945145 +0200
@@ -1404,10 +1404,17 @@ out:
 		if (s <= 0)
 			r = (s << 2) | RESUME_HOST | (r & RESUME_FLAG_NV);
 		else {
-			/* interrupts now hard-disabled */
+			/*
+	                 * interrupts now hard-disabled. i.e EE=0
+			 * But irqs_disabled relies on PACA entries
+			 * and fixup below enables it.
+			 * So irqs_disabled == false
+			 */
 			kvmppc_fix_ee_before_entry();
+			preempt_disable();
 			kvmppc_load_guest_fp(vcpu);
 			kvmppc_load_guest_altivec(vcpu);
+			preempt_enable_no_resched();
 		}
 	}
 
