diff -rupN a/arch/powerpc/kvm/powerpc.c b/arch/powerpc/kvm/powerpc.c
--- a/arch/powerpc/kvm/powerpc.c	2026-09-20 22:48:15.000000000 +0200
+++ b/arch/powerpc/kvm/powerpc.c	2026-09-24 06:51:44.311940896 +0200
@@ -146,6 +146,19 @@ int kvmppc_prepare_to_enter(struct kvm_v
 			   are back at square 1 */
 			continue;
 		}
+		#ifdef CONFIG_PPC64
+	               if (unlikely(local_paca->irq_happened != PACA_IRQ_HARD_DIS)) {
+	               /*
+	                * An interrupt arrived between the soft and hard
+	                * disable. Replay it and retry guest entry.
+	                */
+	                       local_irq_enable();
+	                       local_irq_disable();
+	                       hard_irq_disable();
+	                       continue;
+	               }
+		#endif
+
 
 		guest_enter_irqoff();
 		return 1;
