diff -rupN a/arch/powerpc/kernel/prom_init.c b/arch/powerpc/kernel/prom_init.c
--- a/arch/powerpc/kernel/prom_init.c	2026-03-29 22:40:00.000000000 +0000
+++ b/arch/powerpc/kernel/prom_init.c	2026-03-30 15:21:20.440176660 +0000
@@ -3110,6 +3110,10 @@ static void __init fixup_device_tree_pas
 	prom_printf("Changing device_type of SB600 node...\n");
 
 	prom_setprop(iob, name, "device_type", "isa", sizeof("isa"));
+	u32 address_cells = 3, size_cells = 2, interrupt_cells = 1;
+	prom_setprop(iob, name, "#address-cells", &address_cells, sizeof(u32));
+	prom_setprop(iob, name, "#size-cells", &size_cells, sizeof(u32));
+	prom_setprop(iob, name, "#interrupt-cells", &interrupt_cells, sizeof(u32));
 }
 
 static void __init fixup_device_tree(void)
