--- a/arch/powerpc/kernel/prom_init.c	2025-02-23 21:32:57.000000000 +0100
+++ b/arch/powerpc/kernel/prom_init.c	2025-03-01 09:35:20.120797273 +0100
@@ -3109,6 +3109,10 @@ static void __init fixup_device_tree_pas
 	prom_printf("Changing device_type of SB600 node...\n");
 
 	prom_setprop(iob, name, "device_type", "isa", sizeof("isa"));
+	u32 address_cells = 3, size_cells =2, interrupt_cells = 1;
+	prom_setprop(iob, name, "#address-cells", &address_cells, sizeof(u32));
+	prom_setprop(iob, name, "#size-cells", &size_cells, sizeof(u32));
+	prom_setprop(iob, name, "#interrupt-cells", &interrupt_cells, sizeof(u32));
 }
 
 static void __init fixup_device_tree(void)
